langkit/templates/python_api/iterator_types_py.mako

```diff
@@ -0,0 +1,87 @@
+## vim: filetype=makopython
+
+<%def name="base_decl()">
+
+class _BaseIterator(object):
+    """
+    Base class for Ada iterator bindings.
+    """
+
+    c_element_type = None
+    """
+    Ctype class for iterator elements.
+    """
+
+    __slots__ = ('c_value',)
+
+    def __init__(self, c_value):
+        self.c_value = c_value
+
+    def __repr__(self):
+        return '<{}>'.format(type(self).__name__)
+
+    def clear(self):
+        self.c_value = None
+
+    def __del__(self):
+        self.dec_ref(self.c_value)
+        self.clear()
+
+    @classmethod
+    def wrap(cls, c_value):
+        res = cls(c_value)
+        # inc_ref?
```

Comment:
I don’t think you need an `inc_ref` here: unless I’m missing something, the only use case for `wrap` when handling a value returned from a C function: in these cases, the function returns an ownership share, and the wrap method is here only to convert this value (and there will be no direct user for this value afterwards). It’s worth adding as a comment IMO.